ray-curator/ray_curator/core/utils.py

```diff
@@ -0,0 +1,378 @@
+
+    # Copy Xenna dashboard json if not already present
+    xenna_dashboard_src = os.path.join(
+        os.path.dirname(__file__), "..", "..", "scripts", "xenna_grafana_dashboard.json"
+    )
+    xenna_dashboard_src = os.path.abspath(xenna_dashboard_src)
+    xenna_dashboard_dst = os.path.join(dashboards_path, "xenna_grafana_dashboard.json")
+    if os.path.isfile(xenna_dashboard_src) and not os.path.isfile(xenna_dashboard_dst):
+        shutil.copy(xenna_dashboard_src, xenna_dashboard_dst)
+    return grafana_ini_path
+
+
+def add_ray_prometheus_metrics_service_discovery(ray_temp_dir: str) -> None:
+    """
+    Add the ray prometheus metrics service discovery to the prometheus config.
+    """
+    # Check if ray_temp_dir exists in DEFAULT_NEMO_CURATOR_METRICS_PATH/prometheus.yml, if not add it
+    prometheus_config_path = os.path.join(
+        DEFAULT_NEMO_CURATOR_METRICS_PATH,
+        "prometheus.yml",
+    )
+    with open(prometheus_config_path) as prometheus_config_file:
+        prometheus_config = yaml.safe_load(prometheus_config_file)
+    ray_prom_metrics_service_discovery_path = os.path.join(ray_temp_dir, "prom_metrics_service_discovery.json")
+    if (
+        ray_prom_metrics_service_discovery_path
+        not in prometheus_config["scrape_configs"][0]["file_sd_configs"][0]["files"]
+    ):
+        prometheus_config["scrape_configs"][0]["file_sd_configs"][0]["files"].append(
+            ray_prom_metrics_service_discovery_path
+        )
+        with open(prometheus_config_path, "w") as f:
+            yaml.dump(prometheus_config, f)
+    # Get prometheus port
+    prometheus_port = get_prometheus_port()
+    # Send a curl to prometheus for reloading
+    requests.post(f"http://localhost:{prometheus_port}/-/reload", timeout=5)
+
+
+def is_prometheus_running() -> bool:
+    return any(proc.info["name"].lower() == "prometheus" for proc in psutil.process_iter(["name"]))
+
+
+def is_grafana_running() -> bool:
+    return any(proc.info["name"].lower() == "grafana" for proc in psutil.process_iter(["name"]))
+
+
+def get_next_free_port(start_port: int) -> int:
+    for port in range(start_port, 65535):
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            # SO_REUSEADDR to avoid TIME_WAIT issues on some OSes
+            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            try:
+                s.bind(("localhost", port))
+                # If bind succeeds, port is free
+                return port  # noqa: TRY300
+            except OSError:
+                continue
+    msg = f"No free port found between {start_port} and 65535"
+    raise RuntimeError(msg)
+
+
+def _logger_custom_serializer(
+    _: "loguru.Logger",
+) -> None:
+    return None
+
+
+def _logger_custom_deserializer(
+    _: None,
+) -> "loguru.Logger":
+    # Initialize a default logger
+    return logger
+
+
+def init_or_connect_to_cluster(  # noqa: PLR0913
```

Comment:
Could you point me to where we  connect to the cluster here, seems like we only start it using CLI ?